gnovm/stdlibs/std/emit_event.go

```diff
@@ -0,0 +1,71 @@
+package std
+
+// ref: https://github.com/gnolang/gno/pull/575
+// ref: https://github.com/gnolang/gno/pull/1833
+
+import (
+	"encoding/json"
+	"errors"
+
+	gno "github.com/gnolang/gno/gnovm/pkg/gnolang"
+	abci "github.com/gnolang/gno/tm2/pkg/bft/abci/types"
+)
+
+func X_emit(m *gno.Machine, typ string, attrs []string) {
+	eventAttrs, err := attrKeysAndValues(attrs)
+	if err != nil {
+		m.Panic(typedString(err.Error()))
+	}
+
+	pkgPath := CurrentRealmPath(m)
+	fnIdent := getPrevFunctionNameFromTarget(m, "Emit")
+
+	evt := newGnoEvent(typ, pkgPath, fnIdent, eventAttrs...)
+	ctx := m.Context.(ExecContext)
+	ctx.EventLogger.EmitEvent(evt)
+
+	str, err := json.Marshal(evt)
+	if err != nil {
+		m.Panic(typedString(err.Error()))
+	}
+
+	ctx.EmittedEvents = abci.EventString(str)
+}
+
+func attrKeysAndValues(attrs []string) ([]gnoEventAttribute, error) {
+	attrLen := len(attrs)
+	if attrLen%2 != 0 {
+		return nil, errors.New("cannot pair attributes due to odd count")
+	}
+	eventAttrs := make([]gnoEventAttribute, attrLen/2)
+	for i := 0; i < attrLen-1; i += 2 {
+		eventAttrs[i/2] = gnoEventAttribute{
+			Key:   attrs[i],
+			Value: attrs[i+1],
+		}
+	}
+	return eventAttrs, nil
+}
+
+type gnoEvent struct {
+	Type       string              `json:"type"`
+	PkgPath    string              `json:"pkg_path"`
+	Identifier string              `json:"identifier"`
```

Comment:
Maybe rename this to `Function` or `Emitter function` for clarity?

Reply:
I used the term `ientifier` to strictly refer to the meaning. If I were to change it, `EmitterFunction` seems more approprate.

Reply:
+1, renaming to a more explicit and future-proof name is a good idea. Let's go ahead with "Func."

No need to prefix "Emit" as it's an event, not a call.

Reply:
changed https://github.com/gnolang/gno/pull/1653/commits/7e391a90fd7a58fa2977df725cfd56a24f2b3f77